micropipenv.py

```diff
@@ -709,8 +709,27 @@ def _poetry2pipfile_lock(
         extra_dependencies = set()
 
         for dependency_name, dependency_info in entry.get("dependencies", {}).items():
-            if isinstance(dependency_info, dict) and dependency_info.get("optional", False):
-                extra_dependencies.add(dependency_name)
+            if isinstance(dependency_info, dict):
+                if dependency_info.get("optional", False):
+                    extra_dependencies.add(dependency_name)
+
+                # If the dependency has some markers and it's not specified
+                # as a direct dependency, we should move the markers to the
+                # main dependency definition.
+                if "markers" in dependency_info:
+                    # Is the dependency in default of develop packages?
+                    if dependency_name in default and dependency_name not in pyproject_poetry_section["dependencies"]:
+                        category = default
+                    elif dependency_name in develop and dependency_name not in pyproject_poetry_section["dev-dependencies"]:
+                        category = develop
+                    else:
+                        continue
```

Comment:
Unfortunately, there's a bug with that: the sub-dependency with the marker may be defined in `poetry.lock` after the main dependency. As a result, `dependency_name` will be neither in `default` nor in `develop` at that point of the loop.

I have verified this with `Twisted` and its dependency on `twisted-iocpsupport` (Windows-only).

The only solution I can think of would be to accumulate markers in separate dict and write them back to the dependencies after this loop. But you may think of a better solution.

Reply:
Thanks a lot for the feedback. I've reimplemented it so now it should be order-independent. Could you please send me a link to the poetry.lock of Twisted so I can use it for testing?